File: research/shared/tooling/audio-stage-probes/ima_clipped_scan.py
# SPDX-License-Identifier: Apache-2.0
import pathlib,sys,json,subprocess,array,struct,time,statistics,random,hashlib
from ima_wav_logical import parse,decode,advance,STEP
p=pathlib.Path(sys.argv[1]);p.mkdir(parents=True,exist_ok=True);base=pathlib.Path('research/shared/runs/20260919T232859Z-ima-fixture');exe=p/'scan';table=p/'steps.s32';table.write_bytes(array.array('i',STEP).tobytes());build=['c++','-std=c++17','-O2','-pthread','research/shared/tooling/audio-stage-probes/ima_clipped_scan.cpp','-o',str(exe)];subprocess.run(build,check=True)
(p/'protocol.json').write_text(json.dumps({'scope':'Actual IMA WAV mono/stereo nibble parsing; two associative clipped-transition scans: step-index then predictor. Work-efficient four-chunk local-prefix/carry/correction scan in compiledC++,actual4threads perphase; current per-shift delta rule matchesFFmpeg8.1.2, older product-rounding rule explicitly excluded.','correctness':'Compiled serial/parallel outputs exactly matchFFmpeg,25synthetic controls startingindices0/1/44/87/88 and predictors min/max/zero, nibbleextremes. Invalidheader/code guards.','performance':'Five alternating complete mono+stereo jobs source read/hash/WAVparse/nibble serialization/compiled process/tableload/threadcreation/shutdown/output read/reinterleave versus same compiled serial ordinary decode. <=0.9cost; intermediate storage reported, compilation excluded.'},indent=2))
def pack(parsed):
 records=[]
 for block in parsed['blocks']:
  for (pred,index),codes in zip(block['headers'],block['codes']):records.append(struct.pack('<iiI',pred,index,len(codes))+bytes(codes))
 return struct.pack('<I',len(records))+b''.join(records)
def run(mode,packed,name):
 inp=p/(name+'.bin');inp.write_bytes(packed);out=p/(name+'-'+mode+'.s16');subprocess.run([str(exe),mode,str(table),str(inp),str(out)],check=True);a=array.array('h');a.frombytes(out.read_bytes());return a
def job(mode,ch):
 f=base/f'{ch}ch.wav';raw=f.read_bytes();hashlib.sha256(raw).digest();parsed=parse(raw);a=run(mode,pack(parsed),f'{ch}ch');n=parsed['samplesPerBlock'];out=array.array('h')
 for block in range(len(parsed['blocks'])):
  for i in range(n):
   for c in range(ch):out.append(a[(block*ch+c)*n+i])
 return out
rows=[];refs={}
for ch in[1,2]:
 ref=array.array('h');ref.frombytes(subprocess.check_output(['ffmpeg','-v','error','-i',str(base/f'{ch}ch.wav'),'-f','s16le','-'],stderr=subprocess.DEVNULL));refs[ch]=ref
 for mode in['serial','parallel']:assert job(mode,ch)==ref
 rows.append({'channels':ch,'frames':len(ref)//ch,'serialParallelHostExact':True})
synthetic=[]
for idx in[0,1,44,87,88]:
 for pred in[-32768,-32767,0,32766,32767]:
  codes=([7]*32+[15]*64+[0,8,1,9,2,10,3,11]*16);state=(pred,idx);expected=[pred]
  for code in codes:state=advance(*state,code);expected.append(state[0])
  packed=struct.pack('<IiiI',1,pred,idx,len(codes))+bytes(codes);out=run('parallel',packed,'synthetic');assert list(out)==expected;synthetic.append({'index':idx,'predictor':pred,'samples':len(expected),'exact':True})
bad=p/'invalid.bin';bad.write_bytes(struct.pack('<IiiI',1,0,89,1)+b'\0');r=subprocess.run([str(exe),'parallel',str(table),str(bad),str(p/'invalid-output.s16')]);assert r.returncode==4 and not(p/'invalid-output.s16').exists();(p/'results.json').write_text(json.dumps({'rows':rows,'synthetic':synthetic,'invalidIndexReject':True,'noInvalidOutput':True,'workerThreadsPerScanPhase':4,'scanPhasesPerChannelBlock':4,'deltaRule':'per-shift normative/currentFFmpeg8.1.2; not older product rounding'},indent=2));times={'parallel':[],'serial':[]}
for trial in range(6):
 for mode in (['parallel','serial']if trial%2==0 else ['serial','parallel']):
  start=time.perf_counter_ns();outputs=[job(mode,ch)for ch in[1,2]];ms=(time.perf_counter_ns()-start)/1e6;assert all(out==refs[ch]for out,ch in zip(outputs,[1,2]))
  if trial:times[mode].append(ms)
med={k:statistics.median(v)for k,v in times.items()};ratio=med['parallel']/med['serial'];cost={'timesMS':times,'medianMS':med,'ratio':ratio,'passed':ratio<=.9};(p/'cost-results.json').write_text(json.dumps(cost,indent=2));(p/'commands.log').write_text(json.dumps(build)+'\npython3 research/shared/tooling/audio-stage-probes/ima_clipped_scan.py '+str(p)+'\n');print(cost)

File: research/shared/tooling/audio-stage-probes/ima_clipped_scan.cpp
// SPDX-License-Identifier: Apache-2.0
#include <vector>
#include <array>
#include <fstream>
#include <thread>
#include <cstdint>
#include <string>
#include <algorithm>
struct T{int64_t a,lo,hi;};
int64_t clip(int64_t x,int64_t lo,int64_t hi){return std::max(lo,std::min(hi,x));}
T compose(T x,T y){return {x.a+y.a,clip(x.lo+y.a,y.lo,y.hi),clip(x.hi+y.a,y.lo,y.hi)};}
std::vector<int> scan(const std::vector<int>&delta,int initial,int lo,int hi){size_t n=delta.size();std::vector<T>local(n);std::array<T,4>totals,carry;std::array<std::thread,4>workers;for(size_t t=0;t<4;t++)workers[t]=std::thread([&,t]{T state{0,lo,hi};for(size_t i=n*t/4;i<n*(t+1)/4;i++){state=compose(state,T{delta[i],lo,hi});local[i]=state;}totals[t]=state;});for(auto&w:workers)w.join();T state{0,lo,hi};for(size_t t=0;t<4;t++){carry[t]=state;state=compose(state,totals[t]);}std::vector<int>out(n);for(size_t t=0;t<4;t++)workers[t]=std::thread([&,t]{for(size_t i=n*t/4;i<n*(t+1)/4;i++){T full=compose(carry[t],local[i]);out[i]=clip(initial+full.a,full.lo,full.hi);}});for(auto&w:workers)w.join();return out;}
int main(int argc,char**argv){if(argc!=5)return 2;std::ifstream tb(argv[2],std::ios::binary);std::array<int32_t,89>steps;tb.read((char*)steps.data(),356);if(!tb)return 3;std::ifstream in(argv[3],std::ios::binary);uint32_t count;in.read((char*)&count,4);std::vector<int16_t>out;std::string mode=argv[1];const int adj[8]={-1,-1,-1,-1,2,4,6,8};for(uint32_t b=0;b<count;b++){int32_t pred,index;uint32_t n;in.read((char*)&pred,4);in.read((char*)&index,4);in.read((char*)&n,4);if(!in||index<0||index>88||pred<-32768||pred>32767||n>100000)return 4;std::vector<uint8_t>codes(n);in.read((char*)codes.data(),n);if(!in)return 5;for(auto code:codes)if(code>15)return 6;out.push_back(pred);if(mode=="serial"){for(auto code:codes){int step=steps[index],mag=code&7;int delta=(step>>3)+(mag&4?step:0)+(mag&2?step>>1:0)+(mag&1?step>>2:0);pred=clip(pred+(code&8?-delta:delta),-32768,32767);index=clip(index+adj[mag],0,88);out.push_back(pred);}}else if(mode=="parallel"){std::vector<int>adjust(n),delta(n);for(size_t i=0;i<n;i++)adjust[i]=adj[codes[i]&7];auto indices=scan(adjust,index,0,88);for(size_t i=0;i<n;i++){int step=steps[i?indices[i-1]:index],mag=codes[i]&7;delta[i]=(step>>3)+(mag&4?step:0)+(mag&2?step>>1:0)+(mag&1?step>>2:0);if(codes[i]&8)delta[i]=-delta[i];}auto pcm=scan(delta,pred,-32768,32767);out.insert(out.end(),pcm.begin(),pcm.end());}else return 7;}std::ofstream dst(argv[4],std::ios::binary);dst.write((char*)out.data(),out.size()*2);return dst?0:8;}
